lib/awsinject/awsinject.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2015 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package awsinject injects AWS metadata as labels
+package awsinject
+
+import (
+	"sync"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/services"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/gravitational/trace"
+)
+
+// Tags injects tags as server labels, it is run on
+// Auth Server side, as it can require special IAM role or API keys
+// to query instance tags that are not available as instance metadata
+type Tags struct {
+	sync.Mutex
+	sess     *session.Session
+	services map[string]*ec2.EC2
+}
+
+// NewTags returns new tags injector
+func NewTags() (*Tags, error) {
+	// create an AWS session using default SDK behavior, i.e. it will interpret
+	// the environment and ~/.aws directory just like an AWS CLI tool would:
+	sess, err := session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Tags{
+		sess:     sess,
+		services: make(map[string]*ec2.EC2),
+	}, nil
+}
+
+func (t *Tags) getService(region string) *ec2.EC2 {
+	t.Lock()
+	defer t.Unlock()
+	service, ok := t.services[region]
+	if ok {
+		return service
+	}
+	service = ec2.New(t.sess, aws.NewConfig().WithRegion(region))
+	t.services[region] = service
+	return service
+}
+
+// Inject injects AWS tags for this server
+func (t *Tags) Inject(s services.Server) error {
```

Comment:
This seems to be a blocking call. And it gets called from UpsertServer() which can happen hundreds of times per second. This means that the auth server will be holding hundreds of simultaneous connections to AWS API, so this is code is "hotter" than I'd like. Will there be throttling? What happens if AWS response starts stalling/timing out?

AFAIK `describe-instances` API call can take a list of instance IDs and query their labels at once.